include/fastaq_handler.h

```diff
@@ -4,19 +4,27 @@
 #include <string>
 #include <cstdint>
 #include <fstream>
+#include <exception>
 #include <boost/iostreams/filtering_streambuf.hpp>
 #include <boost/log/core.hpp>
 #include <boost/log/trivial.hpp>
 #include <boost/log/expressions.hpp>
+#include <cstdio>
+#include <zlib.h>
+#include "kseq.h"
+KSEQ_INIT(gzFile, gzread)
 
 namespace logging = boost::log;
 
 struct FastaqHandler {
-    bool gzipped;
-    std::ifstream fastaq_file;
-    boost::iostreams::filtering_istreambuf inbuf;
-    std::istream instream;
-    std::string line;
+private:
+    int closed_status; // see
+                       // https://github.com/attractivechaos/klib/blob/928581a78413bed4efa956731b35b18a638f20f3/kseq.h#L171
+    kseq_t* inbuf;
+
+public:
+    const std::string& filepath;
```

Comment:
Please change to `const std::string filepath;`. I will explain why there is a bug if we leave `const std::string& filepath;`, although it is not a serious bug as we basically don't use the `filepath` variable later (only if there is an error on reading a read from the file... but if in the future we want, for example, to reopen the file, and try to use this variable, we will have bugs). Apologies if this is no novelty to you!

I ran this small test on my machine, that basically creates a `FastaqHandler fh` and prints `fh.filepath` to the screen:
```
TEST(FastaqHandlerTest, filepath_test)
{
    FastaqHandler fh(TEST_CASE_DIR + "reads.fq.gz");
    std::cout << "filepath: " << fh.filepath << std::endl;
}
```

This is the output:
`filepath: �^K��U  st/test_cases/reads.fq.gz`

but should have been:
`filepath: ../../test/test_cases/reads.fq.gz`

There are some corrupted bytes at the start of the string. The bottom line is that we are referring to a memory that is not allocated anymore (i.e. `filepath` refers to a `string` that does not exist anymore), so the process can and might overwrite on that memory, because it sees it as available. This might not be reproduceable in your end, it might output the correct string, or a totally different string, it depends on the execution.

This is why this happens. On this line:
`FastaqHandler fh(TEST_CASE_DIR + "reads.fq.gz");`

c++ creates a _temporary, unnamed_ `std::string` with value `../../test/test_cases/reads.fq.gz` and gives it to the `FastaqHandler` constructor. To simplify, let's call this `string` by `s` (although it is an unnamed variable) and let's say it is at memory address `m`. In the constructor, we assign `filepath <- s`. As `filepath` is a `const std::string&`, we are telling c++ that the `std::string` at memory address `m` now has a new name: `fh.filepath` (`&` in c++ roughly means make alias). So, `fh.filepath` now refers to a `std::string` at memory address `m`. After `FastaqHandler` constructor ends execution, and before the next instruction is executed, c++ sees that the `string` it created at address `m` is just a _temporary, unnamed_  object created just to call the constructor. As we haven't named this variable in the test code (in the context where we created it), its life cycle is just this one line. Thus, c++ cleans up this variable, by calling the destructor on this string and releasing its memory. As the memory occupied by the string is now free, further memory allocations might overwrite the values at memory address `m`. The issue is that the `fh` object is still alive, and `fh.filepath` refers to a `string` that already got destroyed, and thus the bug.

By removing `&` from the variable definition, we tell c++ to make a variable copy, instead of a variable alias. Then, if the original copy gets destroyed later, we don't care, as we have our own local copy.

Not a big issue anyway, as we just use `filepath` in an error logging message, excluding the constructor code, but could become an issue in the future. Hope this was somehow useful... c++ in annoyingly error prone, I think all these issues is avoided in rust with the borrow checker... I am not sure, as I don't know rust, but I am seriously thinking about the switch, because bugs like this always slip through, because the language allows it, and they are hard to find and debug...

Reply:
Wow, that is an awesome explanation! I did not know that. I also have no idea why `filepath` was a reference anyway.